### DISCUSSION TITLE
Cache headers for files

### DISCUSSION THREAD

**Discussion by karniv00l** [pocketbase/pocketbase#808] [Q&A] [answered]
Hello,

since images have unique names, will it be possible to just add cache headers like: `Cache-Control: max-age=31536000`, so they can be cached by CDNs and browsers?

Right now, I'm thinking about leveraging the `OnFileDownloadRequest` hook, though I'm not sure how I can modify headers.

There's always an option to use reverse proxy in case of VPS, but it will be nice to have it built in.

Thanks

**Comment by ganigeorgiev** [comment, marked as answer]
Hm, caching should already be supported since we are utilizing the [`Last-Modified` header](https://github.com/pocketbase/pocketbase/blob/master/tools/filesystem/filesystem.go#L236-L243) (if you are testing locally on Windows this header may not be added automatically because of missing tzdata related to https://github.com/golang/go/issues/38453).

In any case, if you want to register a custom response header you can do something like this:

```go
app.OnFileDownloadRequest().Add(func(e *core.FileDownloadEvent) error {
	e.HttpContext.Response().Header().Set("Cache-Control", "max-age=31536000")

	return nil
})
```

**Reply by karniv00l** [reply]
Thank you!

From my experience only `Cache-Control` works, but I've ended up using Cladflare Cache Rules for this to force cache everything and modify headers at `/api/files/*`

**Reply by ganigeorgiev** [reply]
On second thought after reading [RFC 7232](https://www.rfc-editor.org/rfc/rfc7232), the reason the `Last-Modified` header may not be always reliable is probably because its caching validator is consider "weak" if for example there is no support for the corresponding `If-Modified-Since` request header (there are also other rules that could be found in the "Comparison" section in https://www.rfc-editor.org/rfc/rfc7232#section-2.2.2).

For better reliability, I'll consider adding a default `Cache-Control` header for the file resources in v0.8.

**Reply by karniv00l** [reply]
Awesome!

**Reply by rodydavis** [reply]
> For better reliability, I'll consider adding a default Cache-Control header for the file resources in v0.8.

Did this get added?

Also would love to control the Cache-Control headers for a collection in the admin console too.

There could be collections that only admins can write to but readonly for SDK and Rest API.

By having Cache control headers we could write a simple MPA in pb_public and pb_hooks with libraries with HTMX (which I am exploring)

**Reply by ganigeorgiev** [reply]
@rodydavis 

> Did this get added?

Yes, there should be a default `Cache-Control` header for the uploaded files download endpoint.

There are no plans adding option to controls the header from the Admin UI as it will be too difficult to generalize and for now the hooks are the recommended way of adding/removing user-defined headers.
